generator/src/main/java/com/scottlogic/deg/generator/walker/reductive/ReductiveTreePruner.java

```diff
@@ -0,0 +1,108 @@
+     * @param constraintNode The Tree to be pruned
+     * @param nextFixedField The field, and the value to fix it for.
+     * @return A pruned tree if the new tree is valid, Combined.contradictory otherwise
+     */
+    public Combined<ConstraintNode> pruneConstraintNode(ConstraintNode constraintNode, FixedField nextFixedField) {
+        return pruneConstraintNode(constraintNode, nextFixedField.getField(), nextFixedField.getFieldSpecForCurrentValue());
+    }
+
+    private Combined<ConstraintNode> pruneConstraintNode(ConstraintNode constraintNode, Field field, FieldSpec mergingFieldSpec) {
+        Combined<FieldSpec> newFieldSpec = combineConstraintsWithParent(field, constraintNode, mergingFieldSpec);
+        if (newFieldSpec.isContradictory()){
+            return Combined.contradictory();
+        }
+
+        Collection<AtomicConstraint> newAtomicConstraints = new ArrayList<>(constraintNode.getAtomicConstraints());
+        Collection<DecisionNode> newDecisionNodes = new ArrayList<>();
+
+        for (DecisionNode decision : constraintNode.getDecisions()) {
+            Combined<DecisionNode> prunedDecisionNode = pruneDecisionNode(decision, field, newFieldSpec.get());
+
+            if (prunedDecisionNode.isContradictory()) {
+                return Combined.contradictory();
+            }
+
+            if (onlyOneOption(prunedDecisionNode)) {
+                ConstraintNode remainingConstraintNode = getOnlyRemainingOption(prunedDecisionNode);
+
+                newAtomicConstraints.addAll(remainingConstraintNode.getAtomicConstraints());
+                newDecisionNodes.addAll(remainingConstraintNode.getDecisions());
+            }
+            else {
+                newDecisionNodes.add(prunedDecisionNode.get());
+            }
+        }
+
+        return Combined.of(new TreeConstraintNode(newAtomicConstraints, newDecisionNodes));
+    }
+
+    private Combined<DecisionNode> pruneDecisionNode(DecisionNode decisionNode, Field field, FieldSpec mergingFieldSpec) {
+        Collection<ConstraintNode> newConstraintNodes = new ArrayList<>();
+
+        for (ConstraintNode constraintNode : decisionNode.getOptions()) {
+            pruneConstraintNode(constraintNode, field, mergingFieldSpec).ifPresent(newConstraintNodes::add);
+        }
+
+        if (newConstraintNodes.isEmpty()) {
+            return Combined.contradictory();
+        }
+
+        return Combined.of(new TreeDecisionNode(newConstraintNodes));
+    }
+
+    private Combined<FieldSpec> combineConstraintsWithParent(Field field, ConstraintNode constraintNode, FieldSpec parentFieldSpec){
+        List<AtomicConstraint> atomicConstraintsForField =
+            AtomicConstraintsHelper.getConstraintsForField(constraintNode.getAtomicConstraints(), field);
+
+        Optional<FieldSpec> nodeFieldSpec = constraintReducer.reduceConstraintsToFieldSpec(atomicConstraintsForField);
+
+        if (!nodeFieldSpec.isPresent()) {
```

Comment:
You should check if `atomicConstraintsForField` is empty; i.e. there are no constraints for the given field.

If this case `reduceConstraintsToFieldSpec()` would yield `FieldSpec.Empty`, therefore it will never contradict.